### PR TITLE
[SPARK-25317][CORE] Avoid perf regression in Murmur3 Hash on UTF8String

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

SPARK-10399 introduced a performance regression on the hash computation for UTF8String.

The regression can be evaluated with the code attached in the JIRA. That code runs in about 120 us per method on my laptop (MacBook Pro 2.5 GHz Intel Core i7, RAM 16 GB 1600 MHz DDR3) while the code from branch 2.3 takes on the same machine about 45 us for me. After the PR, the code takes about 45 us on the master branch too.

## How was this patch tested?

running the perf test from the JIRA

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/unsafe/hash/Murmur3_x86_32.java [hunk 1]
```diff
@@ -69,22 +70,27 @@ public static int hashUnsafeWords(Object base, long offset, int lengthInBytes, i
...
-    int lengthInBytes = Ints.checkedCast(base.size());
     assert (lengthInBytes >= 0): "lengthInBytes cannot be negative";
     int lengthAligned = lengthInBytes - lengthInBytes % 4;
-    int h1 = hashBytesByIntBlock(base.subBlock(0, lengthAligned), seed);
+    int h1 = hashBytesByIntBlock(base, lengthAligned, seed);
+    long offset = base.getBaseOffset();
+    Object o = base.getBaseObject();
     for (int i = lengthAligned; i < lengthInBytes; i++) {
-      int halfWord = base.getByte(i);
+      int halfWord = Platform.getByte(o, offset + i);
```
Comment (viirya) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/hash/Murmur3_x86_32.java]: So seems the performance regression is due to the cost of virtual function calls on `MemoryBlock`?

Comment (mgaido91) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/hash/Murmur3_x86_32.java]: that was my guess too at the beginning, but if you just do this change, performance won't change. Seems reasonable what said by @kiszk about the clue being the size of the javabyte code generated, but needs more investigation.

Comment (viirya) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/hash/Murmur3_x86_32.java]: Ok. Seems there are more than single cause for this performance regression.

**Comment by cloud-fan** [comment]
Thanks for working on it! Will it be helpful if we move these hash methods to `MemoryBlock`? e.g. the code can be `int halfWord =bytes[offset + i];`

**Comment by mgaido91** [comment]
@cloud-fan I think I tried doing something like what you suggested but it didn't help. Moreover, the current code in `MemoryBlock` already leverages `Platform.get*` for most of its methods, so that wouldn't really change much I think.

**Comment by cloud-fan** [comment]
This basically reverts the memory block in the hash computing, now the memory block is just a holder of the base object and base offset. This does fix the regression, will we also lose the perf speed up too? IIUC we did observe significant perf boost when introducing the memory block.

**Comment by mgaido91** [comment]
@kiszk yes, I know. Moving `checkedCast` makes a big difference, but if you just move it, without the other changes, there is no perf gain (at least this is what I found in my experiments).

**Comment by kiszk** [comment]
In addition to [your commit](https://github.com/apache/spark/pull/22338/commits/91adce590461dda885d88319a700a775e63f9ce6), I applied the following change, basically use `MemoryBlock` in `hashBytesByIntBlock()`. I got 33us.

```
  private static int hashBytesByIntBlock(MemoryBlock base, int lengthInBytes, int seed) {
    assert (lengthInBytes % 4 == 0);
    int h1 = seed;
    for (int i = 0; i < lengthInBytes; i += 4) {
      int halfWord = base.getInt(i);
      int k1 = mixK1(halfWord);
      h1 = mixH1(h1, k1);
    }
    return h1;
  }
```

But, furthermore, when I also used `MemoryBlock` in `hashUnsafeBytesBlock()`, I got 111us.

**Comment by mgaido91** [comment]
>This does fix the regression, will we also lose the perf speed up too? IIUC we did observe significant perf boost when introducing the memory block.

@cloud-fan I checked the original PR and I couldn't find any benchmark related to Murmur3 hash. The only benchmark I found there was using the `HiveHasher`. So I cannot really answer on this. Do you have a benchmark to test it? Thanks.

**Comment by mgaido91** [comment]
you're right @kiszk. Let me update the PR accordingly in order to narrow down the change/problem, thanks.

**Comment by kiszk** [comment]
While I say these performance differences, I do not understand why these difference occurs completely. That is why I said "I have not found the root cause".
Let us narrow down the problem and find the root cause.

**Comment by mgaido91** [comment]
> While I say these performance differences, I do not understand why these difference occurs completely. That is why I said "I have not found the root cause".

Yes, I am in the same situation. I see the difference but I cannot explain it.

**Comment by viirya** [comment]
Yeah, it's interesting... Seems both `checkedCast` and `Platform.getByte` are changed and performance gets gain. Any single change doesn't work.

**Comment by cloud-fan** [comment]
Since the change looks safer to me and it does fix the regression, I'm merging it to unblock 2.4 release. Please continue to investigate the root cause, thanks!

**Pull request closed without merging**